homeassistant/components/http/__init__.py

```diff
@@ -87,18 +88,54 @@
 STORAGE_VERSION: Final = 1
 SAVE_DELAY: Final = 180
 
+CONF_SERVERS = "servers"
+
+SERVER_SCHEMA_WITHOUT_PORT = {
+    vol.Optional(CONF_SERVER_HOST): vol.All(
+        cv.ensure_list, vol.Length(min=1), [cv.string]
+    ),
+    vol.Optional(CONF_SSL_CERTIFICATE): cv.isfile,
+    vol.Optional(CONF_SSL_PEER_CERTIFICATE): cv.isfile,
+    vol.Optional(CONF_SSL_KEY): cv.isfile,
+    vol.Optional(CONF_SSL_PROFILE, default=SSL_MODERN): vol.In(
+        [SSL_INTERMEDIATE, SSL_MODERN]
+    ),
+}
+
+OPTIONAL_PORT = {vol.Optional(CONF_SERVER_PORT, default=SERVER_PORT): cv.port}
+
+_EXCLUSIVE_PORT_KEY = vol.Exclusive(CONF_SERVER_PORT, "server")
+_EXCLUSIVE_PORT_KEY.default = vol.default_factory(SERVER_PORT)
+
+EXCLUSIVE_PORT = {_EXCLUSIVE_PORT_KEY: cv.port}
+
+
+def _has_all_unique_ports(servers: list[dict[str, Any]]) -> list[dict[str, Any]]:
+    """Validate that each http service has a unique port."""
+    ports = [list[CONF_SERVER_PORT] for list in servers]
+    vol.Schema(vol.Unique())(ports)
+    return servers
+
+
+SERVERS_EXCLUSIVE_MESSAGE = (
+    'Configure one server at top level or configure multiple servers under "servers"'
+)
+
+
 HTTP_SCHEMA: Final = vol.All(
     cv.deprecated(CONF_BASE_URL),
     vol.Schema(
         {
-            vol.Optional(CONF_SERVER_HOST): vol.All(
```

Comment:
Still need to deprecate the old way. 

Would be nice to tell them what to do as well like exclusive has a message